umpf

```diff
@@ -1855,6 +1864,197 @@ do_distribute() {
 	run_distribute
 }
 
+### namespace: push ###
+
+push_topic() {
+	echo "${content}" >> "${STATE}/topic-names"
+}
+
+push_hashinfo() {
+	echo "${content}" >> "${STATE}/topics"
+}
+
+push_release() {
+	echo "${content}" >> "${STATE}/tagname"
+}
+
+push_topic_range() {
+	[ ! -e "${STATE}/tagname" ] && return
+	[ -e "${STATE}/tagrev-flat" ] && abort "more than one 'topic-range' after 'release'!"
+
+	echo "${content##*..}" > "${STATE}/tagrev-flat"
+}
+
+### command: push ###
+
+resolve_commitish() {
+	${GIT} rev-parse --revs-only "$@" 2>/dev/null
+}
+
+shorten_commitish() {
+	resolve_commitish --short ${1}
+}
+
+resolve_tag() {
+	local remote=$1 tag=$2 commit
+	if [ -n "$remote" ]; then
+		# handles conflicting tags on remote
+		commit=$(git ls-remote -q $remote refs/tags/$tag 2>/dev/null | \
+			sed 's/\s\+.*$//')
+	else
+		commit="refs/tags/$tag"
+	fi
+
+	resolve_commitish "${commit}^{}"
+}
+
+update_local() {
+	local success=false args="${1}"
+	local opts
+
+	${FORCE} && opts+="--force"
+
+	local line
+	while read -r line; do
+		local prefix="" suffix=""
+
+		eval set -- ${line}
+		[ ${#} -eq 0 ] && continue
+
+		local refparsed=$(shorten_commitish $3)
+
+		if [ -z "${refparsed}" ]; then
+			prefix=" * [new branch]"
+		elif [[ "${1}" = *~* ]]; then
+		       prefix="   ${1}..$(shorten_commitish ${2})"
+		elif $FORCE; then
+		       prefix=" + ${refparsed}..$(shorten_commitish ${2})"
+		       suffix="  (forced update)"
+		else
+		       prefix=" ! [rejected]"
+		       suffix="  (non-fast-forward)"
+		fi
+
+		printf "%-40s %s -> %s%s\n" "$prefix" $2 $3 "$suffix"
+	done <<< "$args"
+
+	while read -r line; do
+		eval set -- ${line}
+		[ ${#} -eq 0 ] && continue
+
+		if $DRYRUN || git branch $opts $3 $2; then
+			success=true
+		fi
+	done <<< "$args"
+
+	$success || abort
+}
+
+do_push () {
+	local opts args remote
+	local -a branches branch_names
+	local -A topics
+
+	if [ -z "${GIT_REMOTE}" ]; then
+		info "Git remote must be specified. Cannot continue."
+		exit 1
+	fi
+
+	if [ "${GIT_REMOTE}" != "refs/heads/" ]; then
+		remote=${GIT_REMOTE%/}
+	fi
+
+	prepare_persistent push "${@}"
+	parse_series push "${STATE}/series"
+
+	local tagname="$(<"${STATE}/tagname")"
+	local tagrevf="$(<"${STATE}/tagrev-flat")"
+	mapfile -t branches < "${STATE}/topics"
+	mapfile -t branch_names < "${STATE}/topic-names"
+
+	if [ -n "${remote}" ]; then
+		# Needed, so git rev-parse below can check for existent branches
+		git fetch --quiet --no-tags ${remote} 2>/dev/null
```

Comment:
No, use `ls-remote` for the branch checks instead.

That would make a `--force-with-lease` possible as well (I'd like to see that as well at some point but I won't insist on it for this PR).

Reply:
Don't call ls-remote for each branch. Call it once and create a map with branch name as key and hash as value. That should not take longer than any single fetch.

Why do you need to compare the previous commit? Having the same tag for different commits locally and remote is a really bad idea.